### ISSUE TITLE
Couple things that might be useful to have

### ISSUE THREAD

**Issue by masaeedu** [jcpetruzza/barbies#39]
Hello there. I was wondering if you'd be amenable to adding the following classes (and ways of deriving them):

```haskell
type Lens' s a = Lens' { project :: forall f. Functor f => (a -> f a) -> s -> f s }

type ProductLike :: (k -> Type) -> Type
class ProductLike t
  where
  lenses :: t (Lens' (t Identity))
```

and:

```haskell
type BTraversable1 :: (k -> Type) -> Type
class BTraversable t => BTraversable1 t
  where
  btraverse1 :: Apply e => (forall a. f a -> e (g a)) -> b f -> e (b g) 
```

**Comment by jcpetruzza** [comment]
I considered something like `lenses` in the past (maybe for any functor, not just `Identity`?) but I'm not sure I see usages that are not covered well by `generic-lens` already. Do you have scenarios where you'd like to operate on all lenses at once?

For `btraverse1`, would one need a new class? E.g., can't one derive it from `btraverse` and an auxiliary newtype wrapper that pretends an `Apply` is actually an `Applicative`, or something like that?

**Comment by masaeedu** [comment]
> For `btraverse1`, would one need a new class? E.g., can't one derive it from `btraverse` and an auxiliary newtype wrapper that pretends an `Apply` is actually an `Applicative`, or something like that?

@jcpetruzza That makes sense at a high level, but I haven't been able to make it work. I think the relevant newtype is `MaybeApply f a = Either (f a) a` (where `Apply f => Applicative (MaybeApply f)`). But in order to convert a traversal over `MaybeApply f` into a traversal over `f` itself, you end up needing an operation `Apply f => Either (f (b g)) (b g) -> f (b g)`, and I'm honestly not sure how to supply that in general for a "`TraversableB` HKD with at least one position".

You can instead make it work with something like this as the primitive operation:

```haskell
bhead :: b f -> Some b
```

But this is a bit more mysterious for an implementor.

**Comment by jcpetruzza** [comment]
I think I wasn't really very clear, sorry! What I meant is that since "`Apply` is `Applicative` without `pure`", and `btraverse` requires an `Applicative` but doesn't use `pure`, we can use a newtype wrapper to pretend an `Apply` is an `Applicative` whose `pure` will never be called. Something like this: 

```haskell
newtype CantUsePure f a = CantUsePure (f a)
  deriving newtype Functor

instance Apply f => Applicative (CantUsePure f) where
  pure = error "Can't use pure!"
  CantUsePure f <*> CantUsePure x = CantUsePure (f <.> x)

btraverse1
  :: (TraversableB b, Apply e)
  => (forall a . f a -> e (g a))
  -> b f
  -> e (b g)
btraverse1 h
  = coerce . btraverse (CantUsePure . h)
```

Of course, one shouldn't export `CantUsePure`.

I think I rather not introduce a dependency just for this type of functions; and one could put them in a separate package, anyway

**Comment by masaeedu** [comment]
@jcpetruzza Yes, exactly. This is why you would want `Traversable1B` to be a subclass of `TraversableB`.

**Comment by fumieval** [comment]
FYI barbies-th provides an equivalent `AccessorsB` class: https://hackage.haskell.org/package/barbies-th-0.1.9/docs/Barbies-TH.html#t:AccessorsB

It would be great to have this in barbies so that barbies-th focus on the Template Haskell part

**Comment by masaeedu** [comment]
@fumieval FWIW i've revised my position on the "bag of lenses" class. I now believe that what we want is the class of HKDs that are products. A naive encoding of this is pretty straightforward:

```
type HKD k = (k -> Type) -> Type

type HList :: [k] -> HKD k
data HList xs f
  where
  HNil :: HList '[] f
  HCons :: f x -> HList xs f -> HList (x ': xs) f

type ProductB :: HKD k -> Constraint
class ProductB b
  where
  type Components b :: [k]
  to :: b f -> HList (Components b) f
  from :: HList (Components b) f

-- Laws:
-- to . from = id
-- from . to = id
```

I believe this would subsume `ConstraintsB`.
